lib/LLVMIRCodeGen/GlowJIT.cpp

```diff
@@ -160,24 +185,67 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
 #endif
 #endif
       compileLayer_(objectLayer_, SimpleCompiler(TM_)) {
+  //  When passing a null pointer to LoadLibraryPermanently, we request to
+  //  'load' the host process itself, making its exported symbols available for
+  //  execution.
   llvm::sys::DynamicLibrary::LoadLibraryPermanently(nullptr);
 }
 
+GlowJIT::~GlowJIT() {
+  // Run any destructor registered with __cxa_atexit.
+  cxxSymbolOverride_.runDestructors();
+  // Run any destructor discovered in the LLVM IR of the JIT modules.
+  for (auto &dtorRunner : irStaticDestructorRunners_) {
+    cantFail(dtorRunner.runViaLayer(compileLayer_));
+  }
+}
+
 GlowJIT::ModuleHandle GlowJIT::addModule(std::unique_ptr<llvm::Module> M) {
   // Add the set to the JIT with the resolver and a newly created
   // SectionMemoryManager.
+
   auto K = ES_.allocateVModule();
+
+  // Record the static constructors and destructors. We have to do this before
+  // we hand over ownership of the module to the JIT.
+  // Note: This code is based on the LLI/OrcLazyJIT LLVM tool code that is based
+  // on the ORCv1 API (see
+  // https://github.com/llvm-mirror/llvm/blob/release_60/tools/lli/OrcLazyJIT.cpp)
+  // In recent LLVM versions (7+), LLJIT uses the newer ORCv2 API (see
+  // https://github.com/llvm-mirror/llvm/blob/release_70/lib/ExecutionEngine/Orc/LLJIT.cpp).
+  std::vector<std::string> ctorNames, dtorNames;
+  for (auto ctor : orc::getConstructors(*M))
+    ctorNames.push_back(mangle(ctor.Func->getName()));
+  for (auto dtor : orc::getDestructors(*M))
+    dtorNames.push_back(mangle(dtor.Func->getName()));
+
   cantFail(compileLayer_.addModule(K, std::move(M)));
+
+#if LLVM_VERSION_MAJOR == 7
+  CtorDtorRunner<decltype(compileLayer_)> ctorRunner(std::move(ctorNames), K);
+#else
+  LegacyCtorDtorRunner<decltype(compileLayer_)> ctorRunner(std::move(ctorNames),
+                                                           K);
+#endif
+
+  // Run the static constructors and register static destructors.
+  cantFail(ctorRunner.runViaLayer(compileLayer_));
```

Comment:
CI reports this line as the reason for a failure in the DEBUG build.
```
 #8 0x000000000073ddec llvm::orc::GlowJIT::addModule(std::unique_ptr<llvm::Module, std::default_delete<llvm::Module> >) /var/lib/jenkins/workspace/build/../lib/LLVMIRCodeGen/GlowJIT.cpp:232:3
```

Reply:
Yes, that's what I was looking at. I had added `cantFail` when calling `ctorRunner.runViaLayer` to please my IDE, but looking at of the the examples using ORCv1, I see that they don't use `cantFail`: https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/lli/OrcLazyJIT.h#L118

Then, I'll simply remove it to see what is going on.
I'll also integrate you other patch to this PR directly.